### DISCUSSION TITLE
REST API Export IPAM VLAN

### DISCUSSION THREAD

**Discussion by jfargen** [netbox-community/netbox#8432] [Q&A]
Would someone provide a quick howto on how to export from IPAM VLANs using the REST API. I'd like to pull this command down via curl, but haven't had any luck in the inital attempts.

**Comment by candlerb** [comment]
You need to show what you tried, and what happened when you tried it.

For me it's as simple as this:

```
curl https://netbox.example.com/api/ipam/vlans/
```

If that doesn't work, try adding `-v` to the curl line for more info.  Show us the response if it doesn't make sense.

**Reply by jfargen** [reply]
$ curl -X POST -H "Content-Type: application/json" -H "Accept: application/json; indent=4" "http://netbox.local/ipam/vlans/?role=vlan100&export" --data '{
    "username": "aUsername",
    "password": "aGoodPassword",
}': -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 192.168.100.100:80...
* Connected to netbox.local (192.168.100.100) port 80 (#0)
> POST /ipam/vlans/?role=vlan100&export HTTP/1.1
> Host: netbox.local
> User-Agent: curl/7.76.1
> Content-Type: application/json
> Accept: application/json; indent=4
> Content-Length: 59
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< Date: Sat, 22 Jan 2022 15:22:33 GMT
< Server: gunicorn/20.0.4
< Content-Type: text/html
< X-Frame-Options: SAMEORIGIN
< Content-Length: 1386
< Vary: Origin
< 

<!DOCTYPE html>
<html lang="en">
<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <meta name="robots" content="NONE,NOARCHIVE">
  <title>403 Forbidden</title>
  <style type="text/css">
    html * { padding:0; margin:0; }
    body * { padding:10px 20px; }
    body * * { padding:0; }
    body { font:small sans-serif; background:#eee; color:#000; }
    body>div { border-bottom:1px solid #ddd; }
    h1 { font-weight:normal; margin-bottom:.4em; }
    h1 span { font-size:60%; color:#666; font-weight:normal; }
    #info { background:#f6f6f6; }
    #info ul { margin: 0.5em 4em; }
    #info p, #summary p { padding-top:10px; }
    #summary { background: #ffc; }
    #explanation { background:#eee; border-bottom: 0px none; }
  </style>
</head>
<body>
<div id="summary">
  <h1>Forbidden <span>(403)</span></h1>
  <p>CSRF verification failed. Request aborted.</p>


  <p>You are seeing this message because this site requires a CSRF cookie when submitting forms. This cookie is required for security reasons, to ensure that your browser is not being hijacked by third parties.</p>
  <p>If you have configured your browser to disable cookies, please re-enable them, at least for this site, or for &#39;same-origin&#39; requests.</p>

</div>

<div id="explanation">
  <p><small>More information is available with DEBUG=True.</small></p>
</div>

</body>
</html>
* Connection #0 to host netbox.local left intact

**Reply by markkuleinio** [reply]
That doesn't look like a valid API call, `/api` is missing from the URL.

**Reply by candlerb** [reply]
Not only that, it should be a GET not a POST.  The attempt to POST a username and password is just random and won't achieve anything.  Ditto the query parameter `export`.  The parameter `role=vlan100` will only work if you have defined a VLAN role with the exact name `vlan100`.

If Netbox is configured to allow unauthenticated access (`LOGIN_REQUIRED = False`) then the curl I gave before will work, or with pretty-printing:

```
curl -H "Accept: application/json; indent=4" "http://netbox.local/api/ipam/vlans/"
```

If it requires authentication (`LOGIN_REQUIRED = True`) then you will need to go into the GUI admin section, and create a token.  Then you access like this:

```
curl -H "Authorization: Token XXX" -H "Accept: application/json; indent=4" "http://netbox.local/api/ipam/vlans/"
```

as described [here](https://netbox.readthedocs.io/en/stable/rest-api/authentication/).

Once you have this working, then you can extend this with query filters, as per the API documentation.  Maybe `?vid=100` is what you're looking for here.

**Reply by jfargen** [reply]
I do not manage this netbox deployment, just trying to get some data out of it, to do some automation. It doesn't seem my account is enabled to create API tokens and I won't be able to test this until the admin creates and API token.

Thanks for the help.

**Reply by candlerb** [reply]
Interesting.  You're right, the [documentation](https://netbox.readthedocs.io/en/stable/rest-api/authentication/#tokens) does say:

> The creation and modification of API tokens can be restricted per user by an administrator.

But it doesn't say how the restriction is controlled. (Raised separately as #8436)

I just tried it with a brand new unprivileged user - "Active" but no object permissions - and I was able to create a token when logged in as that user (GUI top right-hand corner: (username) > Profile, API tokens, Add a token).  Did you try this already?  What did you see?
